pkg/rbd/snapshot.go

```diff
@@ -0,0 +1,105 @@
+package rbd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ceph/ceph-csi/pkg/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog"
+)
+
+func createRBDClone(ctx context.Context, rbdVol, cloneRbdVol *rbdVolume, cr *util.Credentials) error {
+
+	// generate snapshot from parent volume
+	snap := generateSnapFromVol(rbdVol)
+
+	// update snapshot name as cloned volume name as it will be always unique
+	snap.RbdSnapName = cloneRbdVol.RbdImageName
+
+	// check backend image is present as we are reserving single omap key for
+	// both snapshot and cloned rbd image
+	_, err := getImageInfo(ctx, cloneRbdVol.Monitors, cr, cloneRbdVol.Pool, cloneRbdVol.RbdImageName)
+	if err != nil {
+		if _, ok := err.(ErrImageNotFound); !ok {
+			return status.Errorf(codes.Internal, "failed to get image details for %s", cloneRbdVol.RbdImageName)
+		}
+	} else {
+		// cloned volume already present return success to create snapshot from it
+		return nil
+	}
+
+	// check snapshot is present in parent volume
+	_, err = getSnapInfo(ctx, snap.Monitors, cr, snap.Pool,
+		snap.RbdImageName, snap.RbdSnapName)
+	if err != nil {
+		if _, ok := err.(ErrSnapNotFound); !ok {
+			return status.Error(codes.Internal, err.Error())
+		}
+	} else {
+		goto RESTORE
+	}
+
+	// create snapshot
+	err = createSnapshot(ctx, snap, cr)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create snapshot: %v"), err)
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer func() {
+		if err != nil {
+			if err = deleteSnapshot(ctx, snap, cr); err != nil {
+				klog.Errorf(util.Log(ctx, "failed to cleanup snapshot: %v"), err)
+			}
+		}
+	}()
+RESTORE:
+	// create clone image and delete snapshot
+	err = restoreSnapshot(ctx, cloneRbdVol, snap, cr)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	delErr := deleteSnapshot(ctx, snap, cr)
+	if delErr != nil {
+		klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), delErr)
+		delErr = fmt.Errorf("clone created but failed to delete snapshot due to"+" other failures: %v", delErr)
+		delErr = status.Error(codes.Internal, delErr.Error())
+		return delErr
+	}
+
+	return nil
+}
+
+func cleanUpSnapshot(ctx context.Context, rbdSnap *rbdSnapshot, rbdVol *rbdVolume, cr *util.Credentials) error {
+	err := deleteSnapshot(ctx, rbdSnap, cr)
```

Comment:
Is this the snapshot on the source volume RBD image? Shouldn't that have been deleted in the create step?

Reply:
This is snapshot from cloned volume image

Reply:
... so it should have already been deleted above?

Reply:
we are calling this function if we fail to get snapshot information after creating the snapshot, I think this should be fine

Reply:
... but the actual "RBD" snapshot on the source volume RBD image was already deleted, so what does this do? There isn't an "RBD" snapshot on the CO snapshot RBD image.

Reply:
this is called here https://github.com/ceph/ceph-csi/pull/693/files#diff-35c77cc181d49675cd67a6095520b41fR559 if we fail to get snapshot information (we have deleted the temp snapshot created from source RBD volume but  not the snapshot created from cloned rbd image)

Reply:
Ah, I missed the second "rbd snap create" (`createSnapshot` call) against the new RBD clone image that represents the CO snapshot. Apologies.